linux/ibus-keyman/src/engine.c

```diff
@@ -639,18 +661,25 @@ process_persist_action(
   return TRUE;
 }
 
-static gboolean process_emit_keystroke_action(IBusKeymanEngine *keyman) {
-  if (keyman->char_buffer != NULL) {
-    ibus_keyman_engine_commit_string(keyman, keyman->char_buffer);
-    g_free(keyman->char_buffer);
-    keyman->char_buffer = NULL;
+static gboolean
+process_emit_keystroke_action(IBusKeymanEngine *keyman) {
+  IBusEngine *engine = (IBusEngine *)keyman;
+  if ((engine->client_capabilities & IBUS_CAP_PREFILTER) &&
+      !(engine->client_capabilities & IBUS_CAP_SURROUNDING_TEXT)) {
+    keyman->commit_item->emitting_keystroke = TRUE;
+  } else {
```

Comment:
We could invert this condition and combine it with ll671 and move the ` keyman->commit_item->emitting_keystroke = TRUE;` out of the if statement to avoid repeating it and on ll676?